### ISSUE TITLE
Future is not cancelled when executed in Spring application

### ISSUE THREAD

**Issue by BenediktKersjes** [vavr-io/vavr#2551]
The following code snippet is executed in both a plain Java application and in a Controller class of a Spring application on Tomcat:

```java
Future<Void> f = Future.of(() -> {
    int i = 1;
    while(i < 10){
        try {
            Thread.sleep(200);
        } catch (InterruptedException e) {
        }
        System.out.println(i++);
    }
    return Future.successful(null);
});
Thread.sleep(500);
f.cancel();
try{
    f.get();
}
catch(CancellationException e){
    System.out.println("future cancelled");
}

System.out.println("end");
```

The output for the plain Java application is as follows:
```
1
2
3
future cancelled
end
```
Thus, the thread is interrupted and the future is cancelled after the first three executions of the loop.

However, the output of the Spring application is as follows:
```
1
2
3
future cancelled
end
4
5
6
7
8
9
```
Thus, the future's executor thread continues after the future is cancelled.

Debugging the example shows that the `thread` attribute is `null` when the cancel method on the future is called:
```java
if (mayInterruptIfRunning && this.thread != null) {
  this.thread.interrupt();
}
```

**Comment by danieldietrich** [comment]
Thank you for reporting the issue and also for your detailed investigation!
I will further investigate the issue.

**Comment by danieldietrich** [comment]
Hi again,
I investigated the issue again by studying the Vavr code.
I'm afraid, I don't see any possibility how this could happen. Even if the `ForkJoinPool.commonPool()` size is 1, it can't happen in async mode (which is the case if you call `Future.of`).

Which version of Vavr do you use?

Could you provide a sample repository with a minimal Spring Boot application that shows the error? I would like to debug it.

**Comment by danieldietrich** [comment]
Great, thank you! I will take a look

**Comment by smthelusive** [comment]
I have done some investigation on this and figured: 
- it was reproducing only in Spring because main thread of spring app kept running. The thread running our task is a daemon thread, so in the plain java app it stops because the main thread finishes. If in plain app you add `Thread.sleep(...);` in the end, it will hold the main thread causing the app to behave the same!
- I was curious and tried to use java Futures instead and I reproduced exactly the same behavior. The only difference was that then the extra thread is not a daemon so it would continue running even if we don't hold the main thread. 
- There is a flag `mayInterruptIfRunning` that according to Java documentation:
```mayInterruptIfRunning - true if the thread executing this task should be interrupted; otherwise, in-progress tasks are allowed to complete```
However, when I set it to `true`, it doesn't interrupt the thread. As I searched around, it seems that you're supposed to actively check if an interrupt has been requested and perform cleanup in your thread before stopping it manually. Java documentation is not very explicit about it.

To summarize, it looks like it's not a bug :)

**Comment by danieldietrich** [comment]
Thank you so much Nataliia! That is great to hear.
